### DISCUSSION TITLE
What's the difference between unchanged and unset?

### DISCUSSION THREAD

**Discussion by saintazunya** [SeaQL/sea-orm#770] [Q&A] [answered]
Hi,

I'm new to SeaORM and feels a bit confused when dealing with `ActiveModel`. Thank you for this amazing project.

I don't quite get the point of when should I use [ActiveValue::Unchanged](https://docs.rs/sea-orm/latest/sea_orm/entity/enum.ActiveValue.html#variant.Unchanged) or [ActiveValue::NotSet](https://docs.rs/sea-orm/latest/sea_orm/entity/enum.ActiveValue.html#variant.NotSet) 

I made a small experiment and the returned sql seems to be the same.

### Update 
```rust
assert_eq!(
        Update::one(fruit::ActiveModel {
            id: ActiveValue::set(1),
            name: ActiveValue::set("Orange".to_owned()),
            cake_id: ActiveValue::not_set(),
        })
            .build(DbBackend::Postgres)
            .to_string(),
        Update::one(fruit::ActiveModel {
            id: ActiveValue::set(1),
            name: ActiveValue::set("Orange".to_owned()),
            cake_id: ActiveValue::unchanged(Some(12345)),
        })
            .build(DbBackend::Postgres)
            .to_string(),
    );
```

Both queries returns 
`UPDATE "fruit" SET "name" = 'Orange' WHERE "fruit"."id" = 1`
which means for update there is no difference in `unset` and `unchanged` 
### Insert

```rust
assert_eq!(
        Insert::one(fruit::ActiveModel {
            id: ActiveValue::set(1),
            name: ActiveValue::set("Orange".to_owned()),
            cake_id: ActiveValue::set(Some(12345)),
        })
            .build(DbBackend::Postgres)
            .to_string(),
        Insert::one(fruit::ActiveModel {
            id: ActiveValue::set(1),
            name: ActiveValue::set("Orange".to_owned()),
            cake_id: ActiveValue::unchanged(Some(12345)),
        })
            .build(DbBackend::Postgres)
            .to_string(),
    );
```
Both queries returns 
`INSERT INTO "fruit" ("id", "name", "cake_id") VALUES (1, 'Orange', 12345)`
which means for insert there is no difference in `set` and `unchanged`.

It seems to me that unchanged is completely replaceable, and it makes me confused when to use `unchanged` and when to use `set` / `unset`

**Comment by billy1624** [comment, marked as answer]
Hey @saintazunya, welcome and thanks for the supports!

Generally speaking you don't need `ActiveValue::unchanged( ... )`. All you need is just `set()` and `not_set()`.

For update, what you want is just update the changed / setted values. That's why the unchanged value is being ignored.

For insert, we want all setted values including those unchanged values to be inserted into db.

That's the behaviour for update & insert :)
